### PR TITLE
Update translog writeLocation for flushListener after commit

### PR DESCRIPTION
**Pull request by ywangd**
During flush, the shard can keep processing concurrent indexing requests which advance the translog location. If all changes are then commited by the indexWriter, we will update the `flushListener` with a more up-to-date translog location that is read after the commit.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 1]
```diff
@@ -2223,6 +2223,14 @@ protected void flushHoldingLock(boolean force, boolean waitIfOngoing, ActionList
                     // we need to refresh in order to clear older version values
                     refresh("version_table_flush", SearcherScope.INTERNAL, true);
                     translog.trimUnreferencedReaders();
+                    // This writeLocation read can piggyback on the above trimUnreferencedReaders call to save a readlock
+                    final Translog.Location writeLocationAfterFlush = translog.getLastWriteLocation();
+                    // Update the translog location for flushListener if the writeLocation has changed during the flush and
+                    // indexWriter has committed all the changes. If the indexWriter has uncommitted changes, they will
+                    // be flushed by the next flush as intended.
+                    if (writeLocationAfterFlush.equals(commitLocation) == false && hasUncommittedChanges() == false) {
+                        commitLocation = writeLocationAfterFlush;
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:2232]: Can we assert following:
```
assert writeLocationAfterFlush.compareTo(commitLocation) > 0;
```
?

**Review by henningandersen — COMMENTED** [review]
This seems safe to me too. Will you make it ready for review (adding a test seems prudent)?

**Comment by ywangd** [comment]
Added test and assertion. This PR is now ready for review. Thanks!

**Comment by ywangd** [comment]
@elasticmachine update branch

**Review by henningandersen — APPROVED** [review]
LGTM.

**Pull request merged**